archaius-core/src/main/java/com/netflix/config/ConcurrentCompositeConfiguration.java

```diff
@@ -115,9 +115,16 @@ public class ConcurrentCompositeConfiguration extends ConcurrentMapConfiguration
     private final Set<String> usedProperties = ConcurrentHashMap.newKeySet();
 
     public Set<String> getUsedProperties() {
-        return usedProperties;
+        return Collections.unmodifiableSet(new HashSet<>(usedProperties));
     }
 
+    public Set<String> getAndClearUsedProperties() {
+        synchronized (usedProperties) {
```

Comment:
Synchronizing on the map won't keep it from being modified concurrently by other threads, so there still a race condition where threadA calls getUsedProperties, threadB records a new usage, threadA clears the map, and now threadB's usage is now lost.
The only way to avoid that would be to synchronize all methods that record access, in addition to this one. Not sure that'd be worth the performance hit, but in that case you might as well change the map back to a regular HashMap, instead of paying the ConcurrentHashMap tax.

Reply:
Per offline discussion, updated this to use an AtomicReference instead